### PR TITLE
Applier constructor and run: builder

### PR DESCRIPTION
**Pull request by ash2k**
This is an alternative approach to https://github.com/kubernetes-sigs/cli-utils/pull/454, as suggested in https://github.com/kubernetes-sigs/cli-utils/pull/454#issuecomment-965812739.

/hold
/kind cleanup

### CODE REVIEW

File: pkg/apply/applier.go [hunk 1]
```diff
@@ -114,192 +92,138 @@ func (a *Applier) prepareObjects(localInv inventory.InventoryInfo, localObjs obj
...
-		o.PrunePropagationPolicy = metav1.DeletePropagationBackground
+	if err != nil {
+		postErrorEvent(options, err)
+		return
 	}
 }
 
-func handleError(eventChannel chan event.Event, err error) {
-	eventChannel <- event.Event{
+func postErrorEvent(cfg *Options, err error) {
```
Comment (karlkfi) [inline pkg/apply/applier.go:215]: i think we use "send" as the verb for events elsewhere

File: pkg/apply/applier_apply_options.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+	// inventoryPolicy defines the inventory policy of apply.
+	inventoryPolicy inventory.InventoryPolicy
+
+	// eventListeners is a list of event listeners that are called in specified order when an event happens.
+	// Listeners are never called concurrently with each other.
+	eventListeners []func(event.Event)
+}
+
+func NewOptions() *Options {
```
Comment (karlkfi) [inline pkg/apply/applier_apply_options.go:60]: Would `NewDefaultOptions` be a better name? or is that implied? I'm on the fence.

File: pkg/apply/applier_apply_options.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	return &Options{
+		pollInterval:           defaultPollInterval,
+		prune:                  true,
+		prunePropagationPolicy: metav1.DeletePropagationBackground,
+	}
+}
+
+func (o *Options) ServerSideOptions(serverSideOptions common.ServerSideOptions) *Options {
```
Comment (karlkfi) [inline pkg/apply/applier_apply_options.go:68]: Why did you decide not to use the `With` prefix here?

File: pkg/apply/applier_builder.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+	client                       dynamic.Interface
+	discoClient                  discovery.CachedDiscoveryInterface
+	mapper                       meta.RESTMapper
+	restConfig                   *rest.Config
+	unstructuredClientForMapping func(*meta.RESTMapping) (resource.RESTClient, error)
+	statusPoller                 poller.Poller
+}
+
+// NewBuilder returns a new Applier builder.
+func NewBuilder() *Builder {
```
Comment (karlkfi) [inline pkg/apply/applier_builder.go:38]: Since we probably need this for the Destroyer in the same package, this should maybe be `NewApplierBuilder`.

File: pkg/apply/applier_builder.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+	return &Builder{
+		// Defaults, if any, go here.
+	}
+}
+
+func (b *Builder) New() (*Applier, error) {
```
Comment (karlkfi) [inline pkg/apply/applier_builder.go:44]: I think `Build()` is the conventional terminus when using the builder pattern.

File: pkg/apply/applier_apply_options.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	o.pruneTimeout = pruneTimeout
+	return o
+}
+
+func (o *Options) InventoryPolicy(inventoryPolicy inventory.InventoryPolicy) *Options {
+	o.inventoryPolicy = inventoryPolicy
+	return o
+}
+
+func (o *Options) EventListener(eventListener ...func(event.Event)) *Options {
```
Comment (karlkfi) [inline pkg/apply/applier_apply_options.go:113]: These event channel changes should probably go in a separate PR. They seem enabled by but not required for the Builder pattern. If you leave the applier.Run the way it was, it should significantly simplify this PR.

I'm not sure how I feel about these EventChannel changes, but it doesn't need to block the other changes.

**Review by karlkfi — CHANGES_REQUESTED** [review]
- I like the way the apply options were changed here.
- I mostly like the applier builder too, with some changes.
- I'm on the fence on whether the event channel changes are an improvement. They can probably be punted to a subsequent PR.
- I made a few inline comments about conventional naming

File: pkg/apply/applier_apply_options.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/cli-utils/pkg/apply/event"
+	"sigs.k8s.io/cli-utils/pkg/common"
+	"sigs.k8s.io/cli-utils/pkg/inventory"
+)
+
+const defaultPollInterval = 2 * time.Second
+
+type Options struct {
```
Comment (karlkfi) [inline pkg/apply/applier_apply_options.go:17]: Given that we're changing how to build this, we might want to rename it to `apply.ApplierOptions` too, to be consistent with the `DestroyerOptions`. I think the applier options came first before the destroyer was added.

Also, do we want to make this a private struct so it has to be constructed with the builder?

File: pkg/apply/applier_apply_options.go (hunk 3 above)
Comment (ash2k) [reply pkg/apply/applier_apply_options.go:68]: `With` was a way to namespace functions when they all were in the same "package name namespace". Not that they are methods, perhaps it's better to drop `With` for brevity?

File: pkg/apply/applier_apply_options.go (hunk 7 above)
Comment (ash2k) [reply pkg/apply/applier_apply_options.go:17]: Makes sense, renamed and made private.

Comment (ash2k) [reply pkg/apply/applier_apply_options.go:17]: Hm, actually, if we make it private it will be difficult to use mocks with it. Mocking library will not be able to refer to the type.

File: pkg/apply/applier_apply_options.go (hunk 2 above)
Comment (ash2k) [reply pkg/apply/applier_apply_options.go]: I think it should be `NewApplierOptions` if we are changing the type name to `ApplierOptions`.

**Comment by ash2k** [comment]
@karlkfi I've made most of the changes you suggested, thanks for the review.

> I'm on the fence on whether the event channel changes are an improvement. They can probably be punted to a subsequent PR.

I'd like to keep this change:
- It allows to create reusable and composable consumers out of the box, without having to build fan-out on top of the event channel. For example, I'd like to refactor printers to be usable as callbacks:
  ```go
  type Printer interface {
          Print(Event)
  }
  func (o *ApplierOptions) PrintEvents(p Printer) *ApplierOptions {
	return o.EventListener(func(e event.Event) {
		p.Print(e)
	})
  }
  ```
- It allows library user to decide whether they want concurrency or not. As seen in tests, quite often consumer blocks while apply operation is running (no concurrency -> simpler API).

File: cmd/apply/cmdapply.go [hunk 8]
```diff
@@ -137,7 +138,10 @@ func (r *ApplyRunner) RunE(cmd *cobra.Command, args []string) error {
 
 	// Run the applier. It will return a channel where we can receive updates
 	// to keep track of progress and any issues.
-	a, err := apply.NewApplier(r.factory, invClient)
+	a, err := apply.NewApplierBuilder().
+		WithFactory(r.factory).
```
Comment (seans3) [inline cmd/apply/cmdapply.go:142]: Should this optional parameter be required?

File: pkg/apply/applier_builder.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+	"sigs.k8s.io/cli-utils/pkg/apply/poller"
+	"sigs.k8s.io/cli-utils/pkg/apply/prune"
+	"sigs.k8s.io/cli-utils/pkg/inventory"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/polling"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ApplierBuilder struct {
+	// factory is only used to retrieve things that have not been provided explicitly.
+	factory                      util.Factory
```
Comment (mortent) [inline pkg/apply/applier_builder.go:27]: Do we need all these options? I'm not even sure which ones are mutually exclusive. I think some of it we might be able to clean up by simplifying things internally. I'm also uncertain if we should let users provide things like the RESTMapper, since we in practice have some requirements on the implementation that isn't obvious.

File: pkg/apply/applier_apply_options.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+// CollectEventsInto is a convenience helper that attaches a listener that appends events to the eventSink slice.
+func (o *ApplierOptions) CollectEventsInto(eventSink *[]event.Event) *ApplierOptions {
+	return o.EventListener(func(e event.Event) {
+		*eventSink = append(*eventSink, e)
+	})
+}
+
+// CollectErrorInto is a convenience helper that captures error from the first event of ErrorType type.
```
Comment (mortent) [inline pkg/apply/applier_apply_options.go:132]: As far as I know, there should only ever be one Error event.

File: pkg/apply/applier_apply_options.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/cli-utils/pkg/apply/event"
+	"sigs.k8s.io/cli-utils/pkg/common"
+	"sigs.k8s.io/cli-utils/pkg/inventory"
+)
+
+const defaultPollInterval = 2 * time.Second
+
+type ApplierOptions struct {
```
Comment (mortent) [inline pkg/apply/applier_apply_options.go:17]: Do we need to change the name to ApplierOptions? It seems a bit verbose given that it is in the apply package.

File: pkg/apply/applier_apply_options.go [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+	// inventoryPolicy defines the inventory policy of apply.
+	inventoryPolicy inventory.InventoryPolicy
+
+	// eventListeners is a list of event listeners that are called in specified order when an event happens.
+	// Listeners are never called concurrently with each other.
+	eventListeners []func(event.Event)
+}
+
+func NewApplierOptions() *ApplierOptions {
```
Comment (mortent) [inline pkg/apply/applier_apply_options.go:60]: Why would we have a pointer to the ApplierOptions? With the ApplierOptions just holding data, it seems unnecessary.

File: pkg/apply/applier.go [hunk 13]
```diff
@@ -114,192 +92,138 @@ func (a *Applier) prepareObjects(localInv inventory.InventoryInfo, localObjs obj
...
+// Apply performs the apply step. This happens synchronously with updates
+// on progress and any errors reported back via callbacks.
 // Cancelling the operation or setting timeout on how long to Wait
 // for it complete can be done with the passed in context.
 // Note: There isn't currently any way to interrupt the operation
 // before all the given resources have been applied to the cluster. Any
 // cancellation or timeout will only affect how long we Wait for the
 // resources to become current.
-func (a *Applier) Run(ctx context.Context, invInfo inventory.InventoryInfo, objects object.UnstructuredSet, options Options) <-chan event.Event {
+func (a *Applier) Apply(ctx context.Context, invInfo inventory.InventoryInfo, objects object.UnstructuredSet, options *ApplierOptions) {
```
Comment (mortent) [inline pkg/apply/applier.go:103]: This makes for a somewhat strange API. If it is synchronous, I would expect the results (at least any errors) be return values rather than having to fetch it through the `ApplierOptions`.

Returning a channel does come with some complexity, but it is a pretty standard golang pattern. And it is pretty easy to write simple wrapper functions to make it do everything that the the approach in this PR allows for. 
I'm not convinced passing pointers to structs and errors into functions on the `ApplierOptions` is any better. Examples [here](https://github.com/kubernetes-sigs/cli-utils/pull/485/files#diff-35e39add2210ba4befff203297cc91784fc697568de88aca4194f4ca2f460b81R30) and [here](https://github.com/kubernetes-sigs/cli-utils/pull/485/files#diff-0367d69a6e70baf91f515166e4e2cc6c16b065323518fd37388e5c774a23a5f9R170).

**Review by mortent — CHANGES_REQUESTED** [review]
This PR seems to combine several changes that are mostly unrelated. It seems like it could be split up into separate PRs.

File: pkg/apply/applier_apply_options.go (hunk 11 above)
Comment (mortent) [reply pkg/apply/applier_apply_options.go:17]: I didn't see the discussion around this until now. I'm ok with the change.

File: pkg/apply/applier.go (hunk 13 above)
Comment (mortent) [reply pkg/apply/applier.go:103]: If we go down this path, it seems like we could let the `Apply` function return an error. The current `ErrorEvent` is considered fatal and should be the last event sent before the channel is closed. Then users don't have to worry about collecting results from the `ApplyOptions` unless they are interested in the events.
I thought I saw this mentioned in https://github.com/kubernetes-sigs/cli-utils/pull/454, but now I can't find it.

File: cmd/apply/cmdapply.go (hunk 8 above)
Comment (ash2k) [reply cmd/apply/cmdapply.go:145]: It is not required if all other things, that are derived from the factory, are provided explicitly. This approach allows user to customize what applier uses without creating their own `Factory` implementation (to inject desired objects).

File: pkg/apply/applier_builder.go (hunk 9 above)
Comment (ash2k) [reply pkg/apply/applier_builder.go:27]: I think it's good to allow users to inject dependencies directly. For those who don't need to do that, they can inject just the factory and nothing else. My main motivator is the desire to inject a custom `rest.Config`/`dynamic.Interface` with a dedicated rate limiter. Other parts of my program should use a separate rate limiter from the applier. I could do it by creating a custom factory, but that is quite unwieldy. It's an example use case, anyway.

> I'm also uncertain if we should let users provide things like the RESTMapper, since we in practice have some requirements on the implementation that isn't obvious.

If there is a requirement, ideally code should fail early if it's not met rather than silently not do what it is supposed to (current behavior). I'm going to fix that in https://github.com/kubernetes-sigs/cli-utils/issues/421 by using code from https://github.com/kubernetes/kubernetes/pull/105623. As part of that I'm going to change `meta.RESTMapper` to `meta.ResettableRESTMapper` here so the requirement is explicit.

File: pkg/apply/applier.go [hunk 14]
```diff
@@ -245,7 +215,7 @@ func (a *Applier) Run(ctx context.Context, invInfo inventory.InventoryInfo, obje
 	return eventChannel
 }
 
-type Options struct {
+type ApplierOptions struct {
```
Comment (ash2k) [inline pkg/apply/applier.go:218]: We have `DestroyerOptions`, so renamed this for consistency as suggested here https://github.com/kubernetes-sigs/cli-utils/pull/485#discussion_r755460760.

File: pkg/apply/applier.go [hunk 15]
```diff
@@ -286,9 +256,9 @@ type Options struct {
 
 // setDefaults set the options to the default values if they
 // have not been provided.
-func setDefaults(o *Options) {
-	if o.PollInterval == time.Duration(0) {
-		o.PollInterval = poller.DefaultPollInterval
+func setDefaults(o *ApplierOptions) {
+	if o.PollInterval == 0 {
+		o.PollInterval = defaultPollInterval
```
Comment (ash2k) [inline pkg/apply/applier.go:261]: `poller.DefaultPollInterval` is not used anywhere else, so brought it here, made private and dropped the dependency on the `poller` package.

**Comment by ash2k** [comment]
Thank you for feedback, everybody. I've just pushed an updated version. Removed the channel->callback change and brought back the options struct for the `Run` method. The only major change in this PR is builder for construction of the applier. Later we'd want to do the same for the destroyer, I suppose.

File: pkg/apply/applier_builder.go [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+		statusPoller:  bx.statusPoller,
+		invClient:     bx.invClient,
+		client:        bx.client,
+		openAPIGetter: bx.discoClient,
+		mapper:        bx.mapper,
+		infoHelper:    info.NewInfoHelper(bx.mapper, bx.unstructuredClientForMapping),
+	}, nil
+}
+
+func (b *ApplierBuilder) finalize() (*ApplierBuilder, error) {
```
Comment (karlkfi) [inline pkg/apply/applier_builder.go:64]: Do these all need to be finalized in the same function? Could we make a getter function for each option and just use those in the Build function to finalize each field individually?

**Review by karlkfi — COMMENTED** [review]
Everything looks good except maybe the finalize function. LMK if the suggested approach sounds feasible.

**Review by karlkfi — CHANGES_REQUESTED** [review]
Please amend the commit messages to have fix or feat prefixes. I want to try to start following https://www.conventionalcommits.org/en/v1.0.0/#summary to make change logs easier to produce. Since this is a breaking change, add a BREAKING CHANGE line at the end with detail for each external facing API change.

File: pkg/apply/applier_builder.go (hunk 16 above)
Comment (ash2k) [reply pkg/apply/applier_builder.go:64]: It can certainly be changed to work this way. I personally prefer the current style because I'm not sure what you are suggesting will look clearer in this particular case. I also like how you can put the cursor on a field and see where it's initialized and where it's used within a single method (to check that a field has been initialized before use).

If you don't feel strongly about it, I suggest we keep it this way.

**Review by karlkfi — APPROVED** [review]
/lgtm

File: pkg/apply/applier_builder.go (hunk 16 above)
Comment (ash2k) [reply pkg/apply/applier_builder.go:64]: We are making a copy on the next line so that the builder is reusable. I thought you were talking about improving readability of the code.

**Comment by ash2k** [comment]
Rebased - no changes, just a conflict. PTAL @karlkfi @mortent

File: pkg/apply/applier_builder.go (hunk 16 above)
Comment (karlkfi) [reply pkg/apply/applier_builder.go:64]: cool cool. i was also talking about readability, but that's not a deal breaker.

**Review by karlkfi — APPROVED** [review]
/lgtm

**Comment by ash2k** [comment]
Rebased once again.

File: cmd/apply/cmdapply.go (hunk 8 above)
Comment (mortent) [reply cmd/apply/cmdapply.go:145]: Yeah, allowing users to provide the different dependencies directly instead of through the `Factory` is quite useful.

**Review by mortent — APPROVED** [review]
/approve

**Comment by mortent** [comment]
@karlkfi Can you lgtm this again so we can get it merged?
